source/Engine.cpp

```diff
@@ -2020,19 +2023,36 @@ void Engine::DoCollisions(Projectile &projectile)
 	}
 	else if(projectile.MissileStrength())
 	{
-		// If the projectile did not hit anything, give the anti-missile systems
-		// a chance to shoot it down.
+		// If the projectile did not hit anything, check for anti-missile
+		// systems in range.
 		for(Ship *ship : hasAntiMissile)
 			if(ship == projectile.Target() || gov->IsEnemy(ship->GetGovernment()))
-				if(ship->FireAntiMissile(projectile, visuals))
+				if(ship->IsInAntiMissileRangeOf(projectile))
 				{
-					projectile.Kill();
-					break;
+					antiMissileTargets[ship].push_back(&projectile);
 				}
 	}
 }
 
 
+// Fire anti-missile shots as appropriate for this ship, prioritizing nearby
+// projectiles over distant ones. Like DoCollisions, this function adds visuals
+// directly to the main visuals list. If this is multi-threaded in the future,
+// that will need to change.
+void Engine::DoAntiMissile(Ship *ship)
+{
+	sort(antiMissileTargets[ship].begin(), antiMissileTargets[ship].end(),
+			[ship](Projectile *i, Projectile *j) -> bool
+			{
+				return i->Position().Distance(ship->Position()) < j->Position().Distance(ship->Position());
+			});
+	for(Projectile *projectile : antiMissileTargets[ship])
+		if(ship->FireAntiMissile(*projectile, visuals))
+			projectile->Kill();
```

Comment:
It's technically changing the behavior to do so, but I'm leaning towards allowing it: in your proposed algorithm, multiple pointers to the same `Projectile` can be handed out to any number of nearby ships. Thus, if one ship kills a projectile, some other ship will still have a (valid) pointer to the object, and will consider firing antimissile at that same `Projectile` too.

The game's original algorithm processed each projectile and, once it was killed, moved onto the next.

I'm fine with it, since the proposed algorithm removes that implicit information sharing that happens on a single frame, making the ship AI just a bit more fallible, too.
So, what we should do is add a comment that states it is intentional that we allow the ships to "waste" their antimissile fire on a just-killed projectile.

Reply:
I could also have ships check that a projectile has lifespan>0 before wasting their antimissiles, then remove the check later if people actually want this change of behavior. (I'm personally more comfortable erring on the side of changing fewer behaviors.)